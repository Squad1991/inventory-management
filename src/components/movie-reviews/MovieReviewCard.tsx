import { css, styled } from '@compiled/react';
import React from 'react';
import Avatar from '~src/components/common/avatar/Avatar';
import Rating from '~src/components/common/movie-rating/MovieRating';
import Box from '~src/components/common/primitive/Box';
import Card from '~src/components/common/primitive/Card';
import Inline from '~src/components/common/primitive/Inline';
import Stack from '~src/components/common/primitive/Stack';
import VerticalTextOverflow from '~src/components/common/primitive/VerticalTextOverflow';
import { timeAgo } from '~src/components/common/utils/date-utils';
import { ComponentPropsWithTestId } from '~src/components/common/utils/types';
import { MovieReview } from '~src/routes/movie/types';

interface MovieReviewCardProps extends ComponentPropsWithTestId {
  review: MovieReview;
}

const StyledMovieReviewCard = styled(Card)({
  display: 'flex',
  flexDirection: 'column',
  gap: '1rem',
  cursor: 'pointer',
  padding: '1rem',
});

const MovieReviewCard: React.FC<MovieReviewCardProps> = props => {
  const { review, testId } = props;

  const openReview = () => {
    window.open(`https://www.themoviedb.org/review/${review.id}`, '_blank');
  };

  return (
    <StyledMovieReviewCard data-testid={testId} onClick={openReview}>
      <Inline css={css({ justifyContent: 'space-between' })}>
        <Inline css={css({ gap: '1rem', alignItems: 'center' })}>
          <Avatar imageUrl={review.author_details.avatar_path} title={review.author} />
          <Stack>
            <Box data-testid="author">{review.author}</Box>
            <Box data-testid="date-updated">{timeAgo(review.updated_at)}</Box>
          </Stack>
        </Inline>
        {review.author_details.rating && (
          <Rating testId="movie-rating" rating={review.author_details.rating} />
        )}
      </Inline>
      <VerticalTextOverflow maxLines={2}>{review.content}</VerticalTextOverflow>
    </StyledMovieReviewCard>
  );
};

export default React.memo(MovieReviewCard);
