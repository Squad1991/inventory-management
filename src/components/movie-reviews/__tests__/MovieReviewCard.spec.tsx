import { fireEvent, render, screen } from '@testing-library/react';
import React from 'react';
import { vi } from 'vitest';
import MovieReviewCard from '../MovieReviewCard';

describe('MovieReviewCard', () => {
  it('renders the movie review card correctly', () => {
    const { container } = render(
      <MovieReviewCard
        review={{
          author: 'John cena',
          author_details: {
            name: 'John Cena',
            username: 'johncena',
            avatar_path: 'dummy1.jpg',
            rating: 5,
          },
          content: 'Dummy content',
          created_at: '2022-01-01',
          id: '2',
          updated_at: '2022-01-11',
          url: 'https://www.themoviedb.org/review/2',
        }}
      />,
    );
    expect(container).toMatchSnapshot();
  });
  it('should open review in new window', () => {
    vi.spyOn(window, 'open');
    const { container } = render(
      <MovieReviewCard
        testId="movie-review-card"
        review={{
          author: 'John cena',
          author_details: {
            name: 'John Cena',
            username: 'johncena',
            avatar_path: 'dummy1.jpg',
            rating: 5,
          },
          content: 'Dummy content',
          created_at: '2022-01-01',
          id: '2',
          updated_at: '2022-01-11',
          url: 'https://www.themoviedb.org/review/2',
        }}
      />,
    );
    fireEvent.click(screen.getByTestId('movie-review-card'));
    expect(window.open).toHaveBeenCalledWith('https://www.themoviedb.org/review/2', '_blank');
  });
});
