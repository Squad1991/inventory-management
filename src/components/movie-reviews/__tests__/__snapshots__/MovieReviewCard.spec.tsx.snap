// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`MovieReviewCard > renders the movie review card correctly 1`] = `
<div>
  <style
    data-cmpld="true"
  >
    ._1e0c1txw{display:flex}._2lx21bp4{flex-direction:column}._zulp1j6v{gap:1rem}._80omtlke{cursor:pointer}._ca0q1j6v{padding-top:1rem}._u5f31j6v{padding-right:1rem}._n3td1j6v{padding-bottom:1rem}._19bv1j6v{padding-left:1rem}
  </style>
  <style
    data-cmpld="true"
  >
    ._19itl2jy{border:1px solid #e0e0e0}._2rko1rrh{border-radius:.2rem}._16qsd8sd{box-shadow:0 0 10px rgba(0,0,0,.1)}
  </style>
  <div
    class="_1e0c1txw _19itl2jy _2rko1rrh _16qsd8sd _2lx21bp4 _zulp1j6v _80omtlke _ca0q1j6v _u5f31j6v _n3td1j6v _19bv1j6v"
  >
    <style
      data-cmpld="true"
    >
      ._1bah1yb4{justify-content:space-between}
    </style>
    <style
      data-cmpld="true"
    >
      ._2lx2vrvc{flex-direction:row}
    </style>
    <div
      class="_1e0c1txw _2lx2vrvc _1bah1yb4"
    >
      <style
        data-cmpld="true"
      >
        ._4cvr1h6o{align-items:center}
      </style>
      <div
        class="_1e0c1txw _2lx2vrvc _zulp1j6v _4cvr1h6o"
      >
        <img
          height="30"
          src="https://image.tmdb.org/t/p/original/dummy1.jpg"
          style="width: 30px; height: 30px; border-radius: 50%;"
          title="John cena"
          width="30"
        />
        <div
          class="_1e0c1txw _2lx21bp4"
        >
          <div
            class="_1e0c1txw"
            data-testid="author"
          >
            John cena
          </div>
          <div
            class="_1e0c1txw"
            data-testid="date-updated"
          >
            2 years ago
          </div>
        </div>
      </div>
      <style
        data-cmpld="true"
      >
        ._zulp1kwk{gap:.5rem}
      </style>
      <div
        class="_1e0c1txw _4cvr1h6o _zulp1kwk"
        data-testid="movie-rating"
      >
        <svg
          color="red"
          fill="currentColor"
          height="1em"
          stroke="currentColor"
          stroke-width="0"
          style="color: red;"
          viewBox="0 0 576 512"
          width="1em"
          xmlns="http://www.w3.org/2000/svg"
        >
          <path
            d="M259.3 17.8L194 150.2 47.9 171.5c-26.2 3.8-36.7 36.1-17.7 54.6l105.7 103-25 145.5c-4.5 26.3 23.2 46 46.4 33.7L288 439.6l130.7 68.7c23.2 12.2 50.9-7.4 46.4-33.7l-25-145.5 105.7-103c19-18.5 8.5-50.8-17.7-54.6L382 150.2 316.7 17.8c-11.7-23.6-45.6-23.9-57.4 0z"
          />
        </svg>
        5 / 10
      </div>
    </div>
    <style
      data-cmpld="true"
    >
      ._1bto1l2s{text-overflow:ellipsis}._1reo15vq{overflow-x:hidden}._18m915vq{overflow-y:hidden}._1e0ccj1k{display:-webkit-box}._1yyj1o1x{-webkit-line-clamp:var(--_nr6rwl)}._sudp1e54{-webkit-box-orient:vertical}
    </style>
    <div
      class="_1bto1l2s _1reo15vq _18m915vq _1e0ccj1k _1yyj1o1x _sudp1e54"
      style="--_nr6rwl: 2;"
    >
      Dummy content
    </div>
  </div>
</div>
`;
